src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -62,20 +62,23 @@ trait InlineRewrite extends PredicateExpansion with InlineErrorChecker {
     */
   private[this] def expandStatements(stmts: Seqn, member: Member, program: Program, cond: String => Boolean): Seqn = {
     val noFoldUnfoldStmts = removeFoldUnfolds(stmts, member, program, cond)
-    ViperStrategy.Slim({
-      case inhale@Inhale(expr) =>
-        val expandedExpr = expandExpression(expr, member, program, cond)
-        inhale.copy(expandedExpr)(pos = inhale.pos, info = inhale.info, errT = inhale.errT)
-      case exhale@Exhale(expr) =>
-        val expandedExpr = expandExpression(expr, member, program, cond)
-        exhale.copy(expandedExpr)(pos = exhale.pos, info = exhale.info, errT = exhale.errT)
-      case assert@Assert(expr) =>
-        val expandedExpr = expandExpression(expr, member, program, cond)
-        assert.copy(expandedExpr)(pos = assert.pos, info = assert.info, errT = assert.errT)
-      case loop@While(_, invs, _) =>
-        val expandedInvs = invs.map(expandExpression(_, member, program, cond))
-        loop.copy(invs = expandedInvs)(pos = loop.pos, info = loop.info, errT = loop.errT)
-    }, Traverse.TopDown).execute[Seqn](noFoldUnfoldStmts)
+    ViperStrategy.CustomContext[Set[String]]({
+      case (inhale@Inhale(expr), ctxt) =>
+        val expandedExpr = expandExpression(expr, member, program, cond, ctxt)
+        (inhale.copy(expandedExpr)(pos = inhale.pos, info = inhale.info, errT = inhale.errT), ctxt)
+      case (exhale@Exhale(expr), ctxt) =>
+        val expandedExpr = expandExpression(expr, member, program, cond, ctxt)
+        (exhale.copy(expandedExpr)(pos = exhale.pos, info = exhale.info, errT = exhale.errT), ctxt)
+      case (assert@Assert(expr), ctxt) =>
+        val expandedExpr = expandExpression(expr, member, program, cond, ctxt)
+        (assert.copy(expandedExpr)(pos = assert.pos, info = assert.info, errT = assert.errT), ctxt)
+      case (loop@While(_, invs, _), ctxt) =>
+        val expandedInvs = invs.map(expandExpression(_, member, program, cond, ctxt))
+        (loop.copy(invs = expandedInvs)(pos = loop.pos, info = loop.info, errT = loop.errT), ctxt)
+      case (seqn@Seqn(_, scopedDecls), ctxt) =>
+        (seqn, ctxt ++ scopedDecls.map(_.name))
+      case (expr: Exp, ctxt) => (removeUnfoldings(expr, cond), ctxt)
```

Comment:
Is this safe? Matching on anything of type `Exp` seems kinda somehow weird

Reply:
Ah you don't call `expandExpression` on it, I see

Reply:
I think so. There's only a few places where `Exp`s appear directly in a statement, and those are either covered in the test file, or part of the out-of-scope things I haven't looked at (magic wands, quantified expressions, etc.).
Besides, a stmt is just a tree that can contain exps, and when we reach one we can :sparkles: trust the natural recursion :sparkles:

Reply:
> trust the natural recursion 

I can't anymore after I FAILED MYSELF with not writing `assert`

I have an active build so let me run this locally and see what happens